src/client-side-encryption/state_machine.ts

```diff
@@ -153,176 +151,125 @@ export class StateMachine {
-        const promises = [];
+          break;
+        }
 
-        let request;
-        while ((request = context.nextKMSRequest())) {
-          promises.push(this.kmsRequest(request));
+        case MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS: {
+          const kmsProviders = await executor.askForKMSCredentials();
+          context.provideKMSProviders(serialize(kmsProviders));
+          break;
         }
 
-        Promise.all(promises)
-          .then(() => {
-            context.finishKMSRequests();
-            this.execute(executor, context, callback);
-          })
-          .catch(err => {
-            callback(err);
-          });
+        case MONGOCRYPT_CTX_NEED_KMS: {
+          const requests = Array.from(this.requests(context));
+          await Promise.all(requests);
 
-        return;
-      }
+          context.finishKMSRequests();
+          break;
+        }
 
-      // terminal states
-      case MONGOCRYPT_CTX_READY: {
-        const finalizedContext = context.finalize();
-        // TODO: Maybe rework the logic here so that instead of doing
-        // the callback here, finalize stores the result, and then
-        // we wait to MONGOCRYPT_CTX_DONE to do the callback
-        // @ts-expect-error finalize can change the state, check for error
-        if (context.state === MONGOCRYPT_CTX_ERROR) {
-          const message = context.status.message || 'Finalization error';
-          callback(new MongoCryptError(message));
-          return;
+        case MONGOCRYPT_CTX_READY: {
+          const finalizedContext = context.finalize();
+          // @ts-expect-error finalize can change the state, check for error
+          if (context.state === MONGOCRYPT_CTX_ERROR) {
+            const message = context.status.message || 'Finalization error';
+            throw new MongoCryptError(message);
+          }
+          result = deserialize(finalizedContext, this.options) as T;
+          break;
         }
-        callback(undefined, deserialize(finalizedContext, this.options) as T);
-        return;
-      }
-      case MONGOCRYPT_CTX_ERROR: {
-        const message = context.status.message;
-        callback(
-          new MongoCryptError(
-            message ??
-              'unidentifiable error in MongoCrypt - received an error status from `libmongocrypt` but received no error message.'
-          )
-        );
-        return;
-      }
 
-      case MONGOCRYPT_CTX_DONE:
-        callback();
-        return;
+        default:
+          throw new MongoCryptError(`Unknown state: ${context.state}`);
+      }
+    }
 
-      default:
-        callback(new MongoCryptError(`Unknown state: ${context.state}`));
-        return;
+    if (context.state === MONGOCRYPT_CTX_ERROR || result == null) {
+      const message = context.status.message;
+      throw new MongoCryptError(
+        message ??
+          'unidentifiable error in MongoCrypt - received an error status from `libmongocrypt` but received no error message.'
```

Comment:
Would it be more beneficial here to print the state like before when there's no message?

Reply:
I think we should do both - since we don't always have logging enabled throwing the error is still useful.